repl/src/metta_shim.rs

```diff
@@ -98,7 +104,25 @@ impl MettaShim {
     pub fn exec(&mut self, line: &str) {
         metta_shim_env!{{
             let mut parser = SExprParser::new(line);
-            self.result = self.metta.run(&mut parser).unwrap();
+            let mut runner_state = self.metta.start_run();
+
+            //We don't want any leftover interrupts to break us this time
+            *SIGNAL_STATE.lock().unwrap() = 0;
```

Comment:
No I didn't see it.  I found the issue just by auditing the code.  The race window is probably only open for a few hundred microseconds.  But it's the kind of thing that would happen given infinite users infinite time.  :-P

Seriously though, race conditions are nothing to take lightly.

Reply:
If signal was received I don't see why it should not be handled?

Reply:
There are two signal handlers at play.  When the repl is waiting at a prompt, ctrl-c will exit the program.  When it's in the MeTTa runner loop, it will halt the MeTTa execution (unless it's hit 3 times, in which case we promptly exit)

But there is a race window between when the MeTTa execution has terminated, but we haven't gotten back to the prompt yet.  It's only a few hundred microseconds long, but without major re-architecture that involves changes to rustyline there will be that window.

This line fixes a potential bug where the MeTTa run loop terminated naturally, but the user hit ctrl-c anyway.  If the user waited a millisecond longer they would end up at the prompt and the ctrl-c would have terminated the app.  If they had issued the ctrl-c a millisecond earlier, it would have broken the runner loop and cleared the state var.

But as it is, the handler sets the state var, but the runner loop doesn't clear the state var.  In that situation, if the user wants out of the app, they can hit ctrl-c again, no different from if their first ctrl-c came 1ms earlier.

But if the user only intended to break the MeTTa loop, and then they go on to issue a new MeTTa statement, the user certainly wouldn't want that MeTTa statement to be aborted.

Reply:
Ok, do you mean rustyline has its own signal handler which will interrupt the process when control flow inside a `rl.readline()`. If so then it make sense but it cannot be seen from code thus it is confusing. I would then make the comment above this line more descriptive:
```
// This is to handle the case when user pressed Ctrl+C just after MeTTa interpreter finished processing. In such a case
// `SIGINT_RECEIVED_COUNT` is incremented but `SIGINT` is not handled by `rustyline` signal handler.
// Thus execution is interrupted just after user enters new code.
```

Reply:
Clarifying comment added in 0d51318